src/GafferScene/PrimitiveVariableTweaks.cpp

```diff
@@ -0,0 +1,769 @@
+			}
+
+		}
+
+		std::string idVarName = idPlug()->getValue();
+		if( idVarName.size() )
+		{
+			Int64VectorDataPtr mappedIdListData = new Int64VectorData();
+			std::vector< int64_t > &mappedIdList = mappedIdListData->writable();
+			mappedIdList.reserve( idList->readable().size() );
+
+			std::unordered_set< int64_t > idSet( idList->readable().begin(), idList->readable().end() );
+
+			auto idVar = inputPrimitive->variables.find( idVarName );
+			if( idVar == inputPrimitive->variables.end() )
+			{
+				throw IECore::Exception( fmt::format( "Id invalid, can't find primitive variable \"{}\".", idVarName ) );
+			}
+
+			if( idVar->second.interpolation != targetInterpolation )
+			{
+				throw IECore::Exception( fmt::format(
+					"Id variable \"{}\" : Interpolation \"{}\" doesn't match specified interpolation \"{}\".",
+					idVarName, interpolationToString( idVar->second.interpolation ), interpolationToString( targetInterpolation )
+				) );
+			}
+
+			if( idVar->second.indices )
+			{
+				throw IECore::Exception( fmt::format( "Id variable \"{}\" is not allowed to be indexed.", idVarName ) );
+			}
+
+			if( const IntVectorData *intIdsData = IECore::runTimeCast<IntVectorData>( idVar->second.data.get() ) )
+			{
+				const std::vector<int> &intIds = intIdsData->readable();
+				for( size_t i = 0; i < intIds.size(); i++ )
+				{
+					if( idSet.count( intIds[i] ) )
+					{
+						mappedIdList.push_back( i );
+					}
+				}
+			}
+			else if( const Int64VectorData *int64IdsData = IECore::runTimeCast<Int64VectorData>( idVar->second.data.get() ) )
+			{
+				const std::vector<int64_t> &intIds = int64IdsData->readable();
+				for( size_t i = 0; i < intIds.size(); i++ )
+				{
+					if( idSet.count( intIds[i] ) )
+					{
+						mappedIdList.push_back( i );
+					}
+				}
+			}
+			else
+			{
+				throw IECore::Exception( fmt::format( "Id invalid, can't find primitive variable \"{}\" of type IntVectorData or type Int64VectorData.", idVarName ) );
+			}
+
+			idList = mappedIdListData;
+		}
+	}
+	else if( selectionMode == SelectionMode::MaskPrimitiveVariable && targetInterpolation != PrimitiveVariable::Invalid )
+	{
+		std::string maskVarName = maskVariablePlug()->getValue();
+		auto maskVar = inputPrimitive->variables.find( maskVarName );
+
+		if( maskVar == inputPrimitive->variables.end() )
+		{
+			throw IECore::Exception( fmt::format( "Can't find mask primitive variable \"{}\".", maskVarName ) );
+		}
+
+		if( maskVar->second.interpolation != targetInterpolation )
+		{
+			throw IECore::Exception( fmt::format(
+				"Mask primitive variable \"{}\" has wrong interpolation \"{}\", expected \"{}\".",
+				maskVarName, interpolationToString( maskVar->second.interpolation ), interpolationToString( targetInterpolation )
+			) );
+		}
+
+		// It would be a bit more efficient to directly use the mask to set elements, but to avoid a combinatorial
+		// increase in the number of code paths, we just convert the mask into a list of ids to be tweaked.
```

Comment:
Question : would it make sense for the mask to be the canonical representation, and to convert the id list into that form? With the code as it is, perhaps that'll be less efficient, but we still have to deal with not tweaking the same ID twice - feels like converting to the mask does that for us reasonably efficiently?